### ISSUE TITLE
Setup fine-tuning script for the models 

### ISSUE THREAD

**Issue by nshmyrev** [alphacep/vosk-api#185]
As in 

https://github.com/daanzu/kaldi-active-grammar/issues/33

https://github.com/gooofy/zamia-speech/issues/106

**Comment by dpny518** [comment]
can you provide the data/loca/dict for this model
http://alphacephei.com/vosk/models/vosk-model-small-en-us-0.3.zip
i'll help you write the script that downloads the dict and this model and fine tunes the data/train folder and outputs vosk-model-small-en-us-new

**Comment by federico-zb** [comment]
Could you also provide it for "vosk-model-small-es-0.3"? Thank you very much.
I'm trying to fine-tune it and after that, I'll document the process.

**Comment by nshmyrev** [comment]
Better implementation here:

https://github.com/aarora8/kaldi2/blob/opensat_oct2020/egs/OpenSAT2020/s5/local/chain/run_finetune_tl.sh

**Comment by lalimili6** [comment]
@nshmyrev 
I think this is a good and new example. what your opinion?
https://github.com/kaldi-asr/kaldi/blob/master/egs/libri_css/s5_mono/local/chain/tuning/run_tdnn_1d_ft.sh

**Comment by Ashutosh1995** [comment]
@nshmyrev I am trying to adapt a trained model on Indian English accent to wake word data. I have set up the dataset as per the KALDI format. I am not able to understand how should I change my paths and give the dataset and model paths in it.

**Comment by Ashutosh1995** [comment]
@nshmyrev could you please provide data/lang, data/local/lang, chain tree-dir for Indian English vosk zip folder ?

**Comment by nshmyrev** [comment]
More straigthforward gist:

https://gist.github.com/daanzu/d29e18abb9e21ccf1cddc8c3e28054ff#file-run_finetune_tdnn_1a_daanzu-sh

**Comment by Ashutosh1995** [comment]
@nshmyrev can you please provide the necessary files needed for finetuning in daanzu's script for the Indian english accent model

**Comment by nshmyrev** [comment]
Another useful link

https://github.com/zhaoyi2/CVTE_chain_model_finetune

**Comment by plfbvr** [comment]
Hi :)

is it planned to complete the documentation on acoustic model finetuning (here: https://alphacephei.com/vosk/adaptation) ? The procedure seems for now very unclear... For example:
- is the finetuning possible only for a few models or it is possible with all of them ?
- how am I supposed to organized files in the model to run finetuning ? because it seems differents as in Kaldi organization and I'm not sure of what I'm doing...

**Comment by LuggerMan** [comment]
Hi again, Everybody is asking for input files to finetune. When will they be released?

P.S. I don`t quite understand. Since help needed since august, but no upload of files you surely currently have.

**Comment by Archan2607** [comment]
Hi I am also working on the fine-tuning part on indian english vosk model. 
Can anyone please quide me with the information of preparing a proper documentation or the steps to follow?

Also, @Ashutosh1995 i read one of your threads on this issue, have you got any success with that. Can you please discuss? 

Thanks

**Comment by LuggerMan** [comment]
So, by #773  

> you need lats with nnet3/align_lats.sh

align_lats.sh takes feats.scp as input, where could i find that?

**Comment by nshmyrev** [comment]
> align_lats.sh takes feats.scp as input, where could i find that?

Feats are created with make_mfcc.sh from the data folder with wav.scp/segments

**Comment by LuggerMan** [comment]
@nshmyrev  so i basically need to extract feats from the data on which the model was trained, am i right?

**Comment by nshmyrev** [comment]
> @nshmyrev so i basically need to extract feats from the data on which the model was trained, am i right?

From adaptation data, you do not need training data

**Comment by LuggerMan** [comment]
> From adaptation data, you do not need training data

Ah, ok, now i get it! Thank you

**Comment by nabil6391** [comment]
@nshmyrev by any chance is any video tutorial to fine tune kaldi or vosk models. It would be great. Thanks

**Comment by vikraman22** [comment]
Hi @nshmyrev  I'm trying to finetune the us-english model. It requires **vosk-model-en-us-0.22-compile/exp/finetune_ali** directory to consist of final.mdl, ali.*.gz and tree file. I have got these  file for  the data with which I'm trying to finetune. But the data previously used to train the model is not available to pubic from alphacep.

I got this data from Kaldi model for which I was trying to train from scratch  from **kaldi/egs/mini_librispeech/s5/exp/mono** directory. Can I actually use the files from this directory or can I use files from other directories such as tri3b, tri2b etc ? **Note: I used the same data to train and also using the same to finetune the US-English model.**

Whether the data used while training model is also required? Also final.mdl is initially only available in **./exp/chain/tdnn/final.mdl**.  Can I use the same for _**./exp/nnet3/tdnn_sp/** and **./exp/finetune_ali/**_ directories

**Comment by Ashutosh1995** [comment]
@Archan2607  apologies for the late reply but I was temporarily involved in the ASR training and couldn't work out the training part completely.

**Comment by nshmyrev** [comment]
@vikraman22 please take a note that we do not have official finetuning tutorial, so that has to be trial and error path. 

For trial and error you'd better ask one question a time and try to solve simple question yourself, there no need to ask me to do simple things.

Your chances to get help increase if you submit a documentation on finetuning and finetuning setup as a pull request to our codebase just like we have part on training.

**Comment by Ratevandr** [comment]
Hi!  I am trying to finetune the model vosk-model-ru-0.22. I use "run_finetune_tdnn_1a_daanzu.sh" script for this, and I am missing files ali.*.gz. How can I generate them? 
I tried using "steps/nnet3/align.sh" script, but got error
`ERROR (apply-cmvn[5.5.1009~1-e4940]:Value():util/kaldi-table-inl.h:164) Failed to load object from /home/shmyrev/kaldi/egs/ac/vosk-model-ru-0.22-compile/mfcc/raw_mfcc_test_sova_devices.1.ark:41 (to suppress this error, add the permissive (p, ) option to the rspecifier.`

**Comment by nshmyrev** [comment]
> How can I generate them?

With steps/nnet3/align.sh

> but got error

There must be earlier error since feature files are missing

**Comment by qp450** [comment]
Hello! I am also trying to run daanzu's finetuning script to finetune the German model vosk-model-de-0.21 and am looking for the ali.\*.gz files. I had a look at steps/nnet3/align.sh, as suggested in the previous response, but if I understand correctly, that script requires the data-dir - as in data/train - to run, which is not present in the downloaded model. Could you provide the ali.\*.gz files or indicate which directory to use as the data-dir? 
Thank you very much in advance!

**Comment by nshmyrev** [comment]
> indicate which directory to use as the data-dir

The one with your audio samples you going to use for fine-tuning

**Comment by qp450** [comment]
Thank you for your reply! I read in [this comment](https://github.com/daanzu/kaldi-active-grammar/issues/33#issuecomment-706549384) of the [finetuning discussion](https://github.com/daanzu/kaldi-active-grammar/issues/33), that if the alignment files are generated from the very small amount of finetuning data, as opposed to the large amount of training data, they might be of far inferior quality. This dependency seems to have been confirmed in [daanzu's reply](https://github.com/daanzu/kaldi-active-grammar/issues/33#issuecomment-706721159), who then provided the alignment files for the english model. This is why I thought that the initial alignment files are necessary.

**Comment by nshmyrev** [comment]
>  they might be of far inferior quality.

No, it is wrong. The alignment is just timestamps of phonemes. It doesn't depend on amount of data.

You do not need original model alignment to finetune.

**Comment by qp450** [comment]
Ok great, thanks very much for your help!

**Comment by feng612266** [comment]
> Hi again, Everybody is asking for input files to finetune. When will they be released?
> 
> P.S. I don`t quite understand. Since help needed since august, but no upload of files you surely currently have.

@LuggerMan hello,Did you succeed in your subsequent training

**Comment by feng612266** [comment]
> Hi I am also working on the fine-tuning part on indian english vosk model. Can anyone please quide me with the information of preparing a proper documentation or the steps to follow?
> 
> Also, [@Ashutosh1995](https://github.com/Ashutosh1995) i read one of your threads on this issue, have you got any success with that. Can you please discuss?
> 
> Thanks

@Archan2607 hello,Did you succeed in your subsequent training

**Comment by feng612266** [comment]
> As in
> 
> [daanzu/kaldi-active-grammar#33](https://github.com/daanzu/kaldi-active-grammar/issues/33)
> 
> [gooofy/zamia-speech#106](https://github.com/gooofy/zamia-speech/issues/106)

@nshmyrev hello,I want to add some custom words to the Indian-English small model(vosk-model-small-en-in-0.4) to test the speech recognition effect，First, I used the language model for training, cursor say it is simple,but later I found missing files and needed to compile a version of the model. However, compiling a version of the model requires an email application. I don't know how long it will take to receive an email. Now I want to try the acoustic model training. I would like to confirm whether the acoustic model training can be carried out directly based on the existing model or if a compile(or update) version of the model is also needed. Thank you

**Comment by feng612266** [comment]
> [@nshmyrev](https://github.com/nshmyrev) could you please provide data/lang, data/local/lang, chain tree-dir for Indian English vosk zip folder ?

@Ashutosh1995 hello,Did you succeed in your subsequent training

**Comment by feng612266** [comment]
> More straigthforward gist:
> 
> https://gist.github.com/daanzu/d29e18abb9e21ccf1cddc8c3e28054ff#file-run_finetune_tdnn_1a_daanzu-sh

@nshmyrev hi,i use this but when stage 3 tip this "./steps/nnet3/align_lats.sh: no such file exp/chain/tdnn_base/tree"
# Stage 3
# Sat Oct 18 11:44:18 CST 2025

📐 生成对齐 lattices...
  ⚠️ 语言模型目录不完整，从基础模型复制
./steps/nnet3/align_lats.sh --cmd run.pl --nj 1 --acoustic-scale 1.0 --scale-opts --transition-scale=1.0 --self-loop-scale=1.0 --online-ivector-dir exp/nnet3_chain/ivectors_train_hires data/train_hires data/lang exp/chain/tdnn_base exp/train_lats
steps/nnet2/check_ivectors_compatible.sh: WARNING: The directories do not contain iVector ID.
steps/nnet2/check_ivectors_compatible.sh: WARNING: That means it's you who's reponsible for keeping
steps/nnet2/check_ivectors_compatible.sh: WARNING: the directories compatible
./steps/nnet3/align_lats.sh: no such file exp/chain/tdnn_base/tree

**Comment by nshmyrev** [comment]
> [@nshmyrev](https://github.com/nshmyrev) hi,i use this but when stage 3 tip this "./steps/nnet3/align_lats.sh: no such file exp/chain/tdnn_base/tree"

tree file is part of compilation package

**Comment by feng612266** [comment]
> > [@nshmyrev](https://github.com/nshmyrev) hi,i use this but when stage 3 tip this "./steps/nnet3/align_lats.sh: no such file exp/chain/tdnn_base/tree"
> 
> tree file is part of compilation package

Okay, thanks, but where can I get the compiled package for the Indian English small model? Should I wait for an email reply?

**Comment by nshmyrev** [comment]
We didn't receive any email from you

**Comment by nshmyrev** [comment]
@feng612266 Ok, now I see your email. It will take couple days for me to prepare the version for you since we don't have that particular update package.

**Comment by feng612266** [comment]
> [@feng612266](https://github.com/feng612266) Ok, now I see your email. It will take couple days for me to prepare the version for you since we don't have that particular update package.

Ok, thank you very much. I first compiled and trained with the English version of the small compile model. I found it elsewhere,in this link https://github.com/alphacep/vosk-api/blob/master/python/example/colab/vosk-adaptation.ipynb ,After you send me the small Indian English compile model, I'll Try

**Comment by nshmyrev** [comment]
> > [@feng612266](https://github.com/feng612266) Ok, now I see your email. It will take couple days for me to prepare the version for you since we don't have that particular update package.
> 
> Ok, thank you very much. I first compiled and trained with the English version of the small compile model. I found it elsewhere,in this link https://github.com/alphacep/vosk-api/blob/master/python/example/colab/vosk-adaptation.ipynb ,After you send me the small Indian English compile model, I'll Try

Great, I think finetuning from en model should work equally well actually.

**Comment by feng612266** [comment]
> > > [@feng612266](https://github.com/feng612266) Ok, now I see your email. It will take couple days for me to prepare the version for you since we don't have that particular update package.
> > 
> > 
> > Ok, thank you very much. I first compiled and trained with the English version of the small compile model. I found it elsewhere,in this link https://github.com/alphacep/vosk-api/blob/master/python/example/colab/vosk-adaptation.ipynb ,After you send me the small Indian English compile model, I'll Try
> 
> Great, I think finetuning from en model should work equally well actually.

ok，thankyou,I refer to this link https://alphacephei.com/vosk/lm, "the Update process”，After conducting language training and testing with the version after language training, it was found to be effective, but the accuracy rate did not improve significantly. If the best results are desired, do both the language model and the acoustic model need to be trained? Should the language model be trained first and then the acoustic model on the basis of the compiled version
